lib/rust/ensogl/core/src/application/frp.rs

```diff
@@ -506,3 +506,536 @@ macro_rules! define_endpoints_emit_alias {
         }
     };
 }
+
+/// Generate a set of structures allowing for nice management of FRP inputs, outputs, and commands.
+/// Supersedes `define_endpoints`, which should no longer be used.
+///
+/// Given the definition:
+///
+/// ```compile_fail
+///     define_endpoints_2! { [GLOBAL_OPTS] <GENERIC_PARAMETERS>
+///         Input { [INPUT_OPTS]
+///             input1 (f32),
+///             input2 (),
+///         }
+///         Output { [OUTPUT_OPTS]
+///             output1 (String),
+///             output2 (bool),
+///             output3 (),
+///         }
+///     }
+/// ```
+///
+/// The `INPUT_OPTS`, `OUTPUT_OPTS`, and `GLOBAL_OPTS` are optional and will be passed directly to
+/// input, output, or both FRP network definitions, respectively. For example, you can add the
+/// `TRACE_ALL` option this way.
+///
+/// The `GENERIC_PARAMETERS` are also optional and makes all structures generated by this macro
+/// parametrized by given generics. The parameters may have trait bounds in the same way as in
+/// struct declarations, _except_ that all bounds must be in parenthesis; for example:
+/// `<Foo:(Clone+Debug), Bar:(Debug+'static)>`.
+///
+/// The macro will create two sets of API: a public API and an private API. The public API is
+/// intended to be used to send events into the API and receive events from the API, while the
+/// private API is intended to process the events sent in through the public API and generate
+/// the events going out though the public API.
+///
+/// Please note that additional fields are added automatically to the API. In particular, an
+/// output `focused(bool)`, and inputs `focus()`, `defocus()`, and `set_focus(bool)` are always
+/// defined and connected. They are mainly used for the shortcut manager to send commands only
+/// to focused GUI elements.
```

Comment:
This doc does not explain the final structure of structs that are generated, neither it explains what are the connections between the structures (private, public interfaces, where profiling is, etc). This doc should be a proper introduction to anyone using this macro and I feel that these docs are one of the most important parts of this PR. We never had good docs here and it always caused a big pain among our team members.

Reply:
Expanded docs. Any further feedback appreciated. Where there specific pain points about usage that we should address here?